input/witnessgen.go

```diff
@@ -262,5 +294,116 @@ func (wt WitnessType) GenWitnessFunc(signer Signer,
+	case WitnessKeyHash:
+		fallthrough
+	case CommitmentNoDelay:
+		return P2WKHWitnessSize, false, nil
+
+	// Outputs on a past commitment transaction that pay directly
+	// to us.
+	case CommitmentTimeLock:
+		return ToLocalTimeoutWitnessSize, false, nil
+
+	// Outgoing second layer HTLC's that have confirmed within the
+	// chain, and the output they produced is now mature enough to
+	// sweep.
+	case HtlcOfferedTimeoutSecondLevel:
+		return ToLocalTimeoutWitnessSize, false, nil
+
+	// Incoming second layer HTLC's that have confirmed within the
+	// chain, and the output they produced is now mature enough to
+	// sweep.
+	case HtlcAcceptedSuccessSecondLevel:
+		return ToLocalTimeoutWitnessSize, false, nil
+
+	// An HTLC on the commitment transaction of the remote party,
+	// that has had its absolute timelock expire.
+	case HtlcOfferedRemoteTimeout:
+		return AcceptedHtlcTimeoutWitnessSize, false, nil
+
+	// An HTLC on the commitment transaction of the remote party,
+	// that can be swept with the preimage.
+	case HtlcAcceptedRemoteSuccess:
+		return OfferedHtlcSuccessWitnessSize, false, nil
+
+	// A nested P2SH input that has a p2wkh witness script. We'll mark this
+	// as nested P2SH so the caller can estimate the weight properly
+	// including the sigScript.
+	case NestedWitnessKeyHash:
+		return P2WKHWitnessSize, true, nil
+
+	// The revocation output on a revoked commitment transaction.
+	case CommitmentRevoke:
+		return ToLocalPenaltyWitnessSize, false, nil
+
+	// The revocation output on a revoked HTLC that we offered to the remote
+	// party.
+	case HtlcOfferedRevoke:
+		return OfferedHtlcPenaltyWitnessSize, false, nil
+
+	// The revocation output on a revoked HTLC that was sent to us.
+	case HtlcAcceptedRevoke:
+		return AcceptedHtlcPenaltyWitnessSize, false, nil
+
+	// The revocation output of a second level output of an HTLC.
+	case HtlcSecondLevelRevoke:
+		return ToLocalPenaltyWitnessSize, false, nil
+	}
+
+	return 0, false, fmt.Errorf("unexpected witness type: %v", wt)
+}
+
+// AddWeightEstimation adds the estimated size of the witness in bytes to the
+// given weight estimator and returns the number of CSVs/CLTVs used by the
+// script.
+//
+// NOTE: This is part of the WitnessType interface.
+func (wt StandardWitnessType) AddWeightEstimation(
+	estimator *TxWeightEstimator) (int, int, error) {
+
+	var (
+		csvCount  = 0
+		cltvCount = 0
+	)
+
+	// For fee estimation purposes, we'll now attempt to obtain an
+	// upper bound on the weight this input will add when fully
+	// populated.
+	size, isNestedP2SH, err := wt.SizeUpperBound()
+	if err != nil {
+		return 0, 0, err
+	}
+
+	// If this is a nested P2SH input, then we'll need to factor in
+	// the additional data push within the sigScript.
+	if isNestedP2SH {
+		estimator.AddNestedP2WSHInput(size)
+	} else {
+		estimator.AddWitnessInput(size)
+	}
+
+	switch wt {
+	case CommitmentTimeLock,
+		HtlcOfferedTimeoutSecondLevel,
+		HtlcAcceptedSuccessSecondLevel:
+		csvCount++
```

Comment:
Instead of the `csvCount++`, this could also just be `return 0, 1, nil`. But really, this is still not what you want probably. What is actually needed is a  `TimeLockType()` method returning `Cltv`, `Csv` or `None`. But this is pre-existing and optional to fix.